### PR TITLE
feat: addInvestorUniquenessClaims, getInvestorUniquenessClaims

### PR DESCRIPTION
**Pull request by sansan**
### JIRA Link 

https://polymesh.atlassian.net/browse/DA-68
https://polymesh.atlassian.net/browse/DA-69

### Changelog / Description 

Adds enpoint to create investor uniqueness claims and to fetch them dor a did

### Checklist - 

- [x] New Feature ?
- [x] Updated swagger annotation (if API structure is changed) ?
- [x] Unit Test (if possible) ?
- [x] Updated the Readme.md (if required) ?

[DA-68]: https://polymesh.atlassian.net/browse/DA-68?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[DA-69]: https://polymesh.atlassian.net/browse/DA-69?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: src/claims/claims.controller.ts [hunk 1]
```diff
@@ -67,4 +68,23 @@ export class ClaimsController {
 
     return handleServiceResult(serviceResult);
   }
+
+  @ApiOperation({
+    summary: 'Add Investor uniqueness Claims to the signing Identity',
+    description: 'This endpoint will add Investor uniqueness Claims to the signing Identity',
```
[the remark this answers is not in the document]
Comment (polymath-eric) [reply src/claims/claims.controller.ts:78]: I think the scope is so you can make a claim about a particular Asset. for example, the "affiliated" claim is only in respect to a particular asset and isn't a statement about the Identity for every Asset.

If the SDK says that, then I think this should match like you have it.

File: src/claims/claims.controller.ts [hunk 2]
```diff
@@ -71,4 +72,23 @@ export class ClaimsController {
...
+  })
+  @ApiTransactionResponse({
+    description: 'Transaction response',
```
Comment (polymath-eric) [inline src/claims/claims.controller.ts:81]: I think we `type` here

File: src/claims/dto/add-investor-uniqueness.dto.ts [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+  scope: ScopeDto;
+
+  cddId: string;
+
+  proof: string | ScopeClaimProof;
+
+  scopeId: string;
+
+  expiry?: Date;
+}
```
Comment (prashantasdeveloper) [inline src/claims/dto/add-investor-uniqueness.dto.ts:22]: Decorators are missing here

File: src/claims/models/investor-uniqueness.model.ts [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+      type: ClaimType.InvestorUniqueness,
+      scope: {
+        type: ScopeType.Identity,
+        value: '0x0600000000000000000000000000000000000000000000000000000000000000',
+      },
+      cddId: '0x0600000000000000000000000000000000000000000000000000000000000000',
+      scopeId: '0x0600000000000000000000000000000000000000000000000000000000000000',
+    },
+  })
+  readonly claim: InvestorUniquenessClaim;
```
Comment (prashantasdeveloper) [inline src/claims/models/investor-uniqueness.model.ts:23]: I know its a bit redundant, but you need to create a corresponding model for `InvestorUniquenessClaim` and you that here. That we we get to describe what each field means. We avoid using interface directly from SDK as return param (FYI there are a few cases left from the past where we have done that but we should avoid it)

File: src/identities/identities.controller.ts [hunk 5]
```diff
@@ -465,4 +466,51 @@ export class IdentitiesController {
...
+  })
+  @ApiArrayResponse(InvestorUniquenessModel, {
+    description: 'List of InvestorUniquenessClaims for the given DID',
+    paginated: false,
+  })
+  @Get(':did/investor-uniqueness-claims')
+  async getInvestorUniquenessClaims(
+    @Param() { did }: DidDto,
+    @Query() { includeExpired }: IncludeExpiredFilterDto
+  ): Promise<InvestorUniquenessModel[]> {
```
Comment (prashantasdeveloper) [inline src/identities/identities.controller.ts:497]:
```suggestion
  ): Promise<ResultSet<InvestorUniquenessModel>> {
```

File: src/identities/identities.controller.ts [hunk 6]
```diff
@@ -465,4 +466,51 @@ export class IdentitiesController {
...
+        new InvestorUniquenessModel({
+          issuedAt,
+          expiry,
+          claim,
+          target,
+          issuer,
+        })
+    );
+
+    return results;
```
Comment (prashantasdeveloper) [inline src/identities/identities.controller.ts:514]:
```suggestion
    return { results };
```
Based on your usage of ApiArrayResponse, I think you want to return the above as resultset

**Review by prashantasdeveloper — CHANGES_REQUESTED** [review]
Needs some cleanup and changes in model class

**Review by polymath-eric — APPROVED** [review]

File: src/claims/dto/add-investor-uniqueness.dto.ts [hunk 7]
```diff
@@ -12,11 +12,23 @@ export class AddInvestorUniquenessDto extends TransactionBaseDto {
   })
   scope: ScopeDto;
 
+  @ApiProperty({
+    description: 'The CDD ID of the investor',
+  })
```
Comment (prashantasdeveloper) [inline src/claims/dto/add-investor-uniqueness.dto.ts:17]: example field is missing. Also all fields should be marked readonly as well along with decorators. Even if there is nothing to add you must add @IsString or somthing like this

File: src/claims/models/investor-uniqueness-claim.model.ts [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+import { FromEntity } from '~/common/decorators/transformation';
+
+export class InvestorUniquenessClaimModel {
+  @ApiProperty({
+    type: 'string',
+    description: 'Claim type',
+    example: 'InvestorUniqueness',
+  })
+  @FromEntity()
+  readonly type: ClaimType.InvestorUniqueness;
```
Comment (prashantasdeveloper) [inline src/claims/models/investor-uniqueness-claim.model.ts:16]: Do we need to explicitly define the type here? IMO the name suggests itself that the claim type is `InvestorUniqueness`. What do you think @polymath-eric ?

File: src/claims/models/investor-uniqueness-claim.model.ts [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+
+  @ApiProperty({
+    type: ScopeModel,
+    description: 'Scope of the Claim',
+  })
+  @FromEntity()
```
Comment (prashantasdeveloper) [inline src/claims/models/investor-uniqueness-claim.model.ts:22]:
```suggestion
  @Type(() => ScopeModel)
```

File: src/claims/models/investor-uniqueness.model.ts [hunk 10]
```diff
@@ -1,24 +1,14 @@
...
-        type: ScopeType.Identity,
-        value: '0x0600000000000000000000000000000000000000000000000000000000000000',
-      },
-      cddId: '0x0600000000000000000000000000000000000000000000000000000000000000',
-      scopeId: '0x0600000000000000000000000000000000000000000000000000000000000000',
-    },
+    type: InvestorUniquenessClaimModel,
   })
-  readonly claim: InvestorUniquenessClaim;
+  @FromEntity()
```
Comment (prashantasdeveloper) [inline src/claims/models/investor-uniqueness.model.ts:12]:
```suggestion
  @Type(() => InvestorUniquenesssClaimModel)
```

**Review by prashantasdeveloper — CHANGES_REQUESTED** [review]

File: src/claims/models/investor-uniqueness.model.ts [hunk 11]
```diff
@@ -0,0 +1,14 @@
...
+import { FromEntity } from '~/common/decorators/transformation';
+
+export class InvestorUniquenessModel extends PartialType(OmitType(ClaimModel, ['claim'] as const)) {
+  @ApiProperty({
+    description: 'Details of the Claim containing type and scope',
+    type: InvestorUniquenessClaimModel,
+  })
+  @FromEntity()
+  readonly claim: InvestorUniquenessClaimModel;
+}
```
Comment (prashantasdeveloper) [inline src/claims/models/investor-uniqueness.model.ts:14]: We can get rid of this class and directly use ClaimModel<InvestorUniquenessClaimModel>, only catch would be that to properly handle the swagger schema. We would probably have to create a custom decorator to handle this I suppose. Thoughts @polymath-eric ?

File: src/claims/models/investor-uniqueness-claim.model.ts (hunk 8 above)
Comment (polymath-eric) [reply src/claims/models/investor-uniqueness-claim.model.ts:16]: I think you are right.

At first I thought the Model name isn't returned to the client, but if we have a unique model for this its redundant. I think we only need `type` if we were reusing the generic claim model.

File: src/common/decorators/swagger.ts [hunk 12]
```diff
@@ -47,7 +49,59 @@ export const ApiArrayResponse = <TModel extends Type | string>(
     extraModels.push(model);
     items = { $ref: getSchemaPath(model) };
   }
+
+  if (Object.keys(extendItems).length > 0) {
```
Comment (polymath-eric) [inline src/common/decorators/swagger.ts:53]: I think it would be better to make another decorator for this use case and hard code stuff for the claim, rather than introduce this complexity into a decorator that is widely used that didn't really need this stuff before.

I find its rather hard to follow what is happening here, it doesn't seem like a simplification from what was in a previous commit to me.

Comment (prashantasdeveloper) [reply src/common/decorators/swagger.ts:53]: I know this works at the moment, but my idea was to make the such a decorator that could handle ClaimModel<T> only. The change that you have made makes it workable with ApiArrayResponse but if we have to make it work individually that wouldn't work. 

Agreeing with Eric here, even if we aren't able to find an optimal solution we should separate this into a separate decorator and use it for our specific instance keeping the ApiArrayResponse untouched.

Comment (prashantasdeveloper) [reply src/common/decorators/swagger.ts:53]: I'll try to take a shot at it from my side and see if I can come up with something, until then I would like you to wait

**Review by polymath-eric — COMMENTED** [review]
looking better

still some smells and console.logs to clean up

**Comment by polymath-eric** [comment]
@sansan FYI the sonar cloud error is code smell for unused import + a TODO comment.

Not sure what the ts-error is, but if you need to use a cast I think thats better than having the `@ts-expect-error` annotation.

File: src/claims/dto/add-investor-uniqueness.dto.ts [hunk 13]
```diff
@@ -0,0 +1,50 @@
...
+    description: 'The scope ID of the claim',
+    example: '0x0600000000000000000000000000000000000000000000000000000000000000',
+  })
+  @IsString()
+  readonly scopeId: string;
+
+  @ApiProperty({
+    description: 'The expiry date of the claim',
+    example: '2020-01-01',
+  })
```
Comment (polymath-eric) [inline src/claims/dto/add-investor-uniqueness.dto.ts:48]:
```suggestion
  })
  @IsOptional()
  @IsDate()
```

File: src/claims/dto/add-investor-uniqueness.dto.ts [hunk 14]
```diff
@@ -0,0 +1,50 @@
...
+import { ScopeClaimProofDto } from '~/claims/dto/scope-claim-proof.dto';
+import { ScopeDto } from '~/claims/dto/scope.dto';
+import { ApiPropertyOneOf } from '~/common/decorators/swagger';
+import { TransactionBaseDto } from '~/common/dto/transaction-base-dto';
+
+@ApiExtraModels(ScopeClaimProofDto)
+export class AddInvestorUniquenessDto extends TransactionBaseDto {
+  @ApiProperty({
+    description: 'The type of Claim. Note that different types require different fields',
+  })
```
Comment (polymath-eric) [inline src/claims/dto/add-investor-uniqueness.dto.ts:16]:
```suggestion
  })
  @Type(() => ScopeDto)
```

File: src/claims/dto/add-investor-uniqueness.dto.ts [hunk 15]
```diff
@@ -0,0 +1,50 @@
...
+  @ApiPropertyOneOf({
+    description: 'The proof of the claim',
+    union: [
+      {
+        type: 'string',
+        example: '0x0600000000000000000000000000000000000000000000000000000000000000',
+      },
+      ScopeClaimProofDto,
+    ],
+  })
```
Comment (polymath-eric) [inline src/claims/dto/add-investor-uniqueness.dto.ts:35]: I think this should get a validation decorator as well

File: src/claims/dto/proof-scope-id-cdd-id-match.dto.ts [hunk 16]
```diff
@@ -0,0 +1,28 @@
...
+  })
+  @IsArray()
+  @ArrayMinSize(2)
+  @ArrayMaxSize(2)
+  @IsString({ each: true })
+  readonly challengeResponses: [string, string];
+
+  @ApiProperty({
+    type: 'string',
+  })
```
Comment (polymath-eric) [inline src/claims/dto/proof-scope-id-cdd-id-match.dto.ts:18]:
```suggestion
  })
  @IsString()
```

File: src/claims/dto/proof-scope-id-cdd-id-match.dto.ts [hunk 17]
```diff
@@ -0,0 +1,28 @@
...
+    required: true,
+  })
+  @IsArray()
+  @ArrayMinSize(2)
+  @ArrayMaxSize(2)
+  @IsString({ each: true })
+  readonly challengeResponses: [string, string];
+
+  @ApiProperty({
+    type: 'string',
```
Comment (polymath-eric) [inline src/claims/dto/proof-scope-id-cdd-id-match.dto.ts:17]: example?

File: src/claims/dto/proof-scope-id-cdd-id-match.dto.ts [hunk 18]
```diff
@@ -0,0 +1,28 @@
...
+  })
+  readonly subtractExpressionsRes: string;
+
+  @ApiProperty({
+    type: 'string',
+    description: 'The blinded scope DID hash',
+    example: '0x060000',
+    required: true,
+  })
```
Comment (polymath-eric) [inline src/claims/dto/proof-scope-id-cdd-id-match.dto.ts:26]:
```suggestion
  })
  @IsString()
```

File: src/claims/dto/scope-claim-proof.dto.ts [hunk 19]
```diff
@@ -0,0 +1,27 @@
...
+  @ApiPropertyOneOf({
+    description: 'The proof scope Id of the claim',
+    union: [
+      {
+        type: 'string',
+        example: '0x0600000000000000000000000000000000000000000000000000000000000000',
+      },
+      ProofScopeIdCddIdMatchDto,
+    ],
+  })
```
Comment (polymath-eric) [inline src/claims/dto/scope-claim-proof.dto.ts:25]: validation?

File: src/claims/models/investor-uniqueness-claim.model.ts [hunk 20]
```diff
@@ -0,0 +1,41 @@
...
+    description: 'Scope of the Claim',
+  })
+  @Type(() => ScopeModel)
+  readonly scope: ScopeModel;
+
+  @ApiProperty({
+    type: 'string',
+    description: 'CDD ID of the Claim',
+    example: '0x0600000000000000000000000000000000000000000000000000000000000000',
+  })
```
Comment (polymath-eric) [inline src/claims/models/investor-uniqueness-claim.model.ts:28]:
```suggestion
  })
  @IsString()
```

File: src/claims/models/investor-uniqueness-claim.model.ts [hunk 21]
```diff
@@ -0,0 +1,41 @@
...
+  readonly cddId: string;
+
+  @ApiProperty({
+    type: 'string',
+    description: 'Scope ID of the Claim',
+    example: '0x0600000000000000000000000000000000000000000000000000000000000000',
+  })
```
Comment (polymath-eric) [inline src/claims/models/investor-uniqueness-claim.model.ts:35]:
```suggestion
  })
  @IsString()
```

**Review by polymath-eric — CHANGES_REQUESTED** [review]
My understanding is DTOs need 2 types of decorators. The Swagger documentation ones, as well as class-validator ones so the user given values line up with the type annotations.

File: src/claims/models/investor-uniqueness-claim.model.ts (hunk 20 above)
Comment (polymath-eric) [reply src/claims/models/investor-uniqueness-claim.model.ts:28]: This is a model, so ignore here

File: src/claims/models/investor-uniqueness-claim.model.ts (hunk 21 above)
Comment (polymath-eric) [reply src/claims/models/investor-uniqueness-claim.model.ts:35]: this is a model, so ignore this comment here

File: src/claims/dto/proof-scope-id-cdd-id-match.dto.ts (hunk 18 above)
Comment (polymath-eric) [reply src/claims/dto/proof-scope-id-cdd-id-match.dto.ts:26]: This is a DTO, so should get validated imo

File: src/claims/dto/proof-scope-id-cdd-id-match.dto.ts (hunk 16 above)
Comment (polymath-eric) [reply src/claims/dto/proof-scope-id-cdd-id-match.dto.ts:18]: This is DTO, so should have validation

File: src/claims/dto/add-investor-uniqueness.dto.ts (hunk 13 above)
Comment (polymath-eric) [reply src/claims/dto/add-investor-uniqueness.dto.ts:48]: Why no validation here? This is a DTO

**Review by polymath-eric — APPROVED** [review]
LGTM

**Review by prashantasdeveloper — APPROVED** [review]
